contracts/libraries/LibIndexMap.sol

```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.9;
+
+struct IndexMap {
+  uint256[] inner;
+}
+
+using LibIndexMap for IndexMap global;
+
+library LibIndexMap {
+  uint256 internal constant MAX_BIT = 256;
+
+  /**
+   * @dev Wraps an array of uint256 values into an IndexMap struct.
+   * @param inner The array of uint256 values to wrap.
+   * @return The wrapped IndexMap struct.
+   */
+  function wrap(uint256[] memory inner) internal pure returns (IndexMap memory) {
+    return IndexMap(inner);
+  }
+
+  /**
+   * @dev Creates a bitmap array based on the given number of elements.
+   * @param numElement The number of elements to create the bitmap for.
+   * @return bitmap The created bitmap array.
+   */
+  function create(uint16 numElement) internal pure returns (IndexMap memory bitmap) {
+    unchecked {
+      bitmap.inner = new uint256[](1 + uint256(numElement) / MAX_BIT);
+    }
+  }
+
+  /**
+   * @dev Checks if an index is present in the map.
+   * @param bitmap The map to check.
+   * @param index The index to check.
+   * @return A boolean indicating whether the index is present in the map.
+   */
+  function contains(IndexMap memory bitmap, uint256 index) internal pure returns (bool) {
+    unchecked {
+      uint256 size = MAX_BIT;
+      // if index is out of range, return false
+      if (index >= bitmap.inner.length * size) return false;
+      return (bitmap.inner[index / size] & (1 << (index % size))) != 0;
+    }
+  }
+
+  /**
+   * @dev Set of an element in a bitmap based on its value.
+   * @param bitmap The bitmap to record the existence of the element.
+   * @param index The value to record.
+   * @return The updated bitmap with recorded existence of the element.
+   * @notice The bitmap is updated in place.
+   */
+  function set(IndexMap memory bitmap, uint256 index) internal pure returns (IndexMap memory) {
+    unchecked {
+      uint256 size = MAX_BIT;
+      bitmap.inner[index / size] |= 1 << (index % size);
+      return bitmap;
+    }
+  }
+
+  /**
+   * @dev Set of elements in a bitmap based on their values.
+   * @param bitmap The bitmap to record the existence of elements.
+   * @param indices The array of indices to record.
+   * @return The updated bitmap with recorded existence of elements.
+   * @notice The bitmap is updated in place.
+   */
+  function setBatch(IndexMap memory bitmap, uint256[] memory indices) internal pure returns (IndexMap memory) {
```

Comment:
why called `set-` while this is a `pure` method?

Reply:
it modifies memory bitmap